### DISCUSSION TITLE
Friend access modifier

### DISCUSSION THREAD

**Discussion by K1vs** [dotnet/csharplang#7578] [Language Ideas]
Hello,

When I design applications using DDD principles, I often encounter the following problem:

Let's imagine what I have an Order aggregate and an OrderItem entity inside it. E.g. this [Order](https://github.com/dotnet-architecture/eShopOnContainers/blob/dev/src/Services/Ordering/Ordering.Domain/AggregatesModel/OrderAggregate/Order.cs) and  this [OrderItem](https://raw.githubusercontent.com/dotnet-architecture/eShopOnContainers/dev/src/Services/Ordering/Ordering.Domain/AggregatesModel/OrderAggregate/OrderItem.cs).
Order has 
`public void AddOrderItem(int productId, string productName, decimal unitPrice, decimal discount, string pictureUrl, int units = 1)` method, 
and OrderItem  has 
`public void AddUnits(int units)` method.

According to DDD rules (see comment of [AddOrderItem](https://raw.githubusercontent.com/dotnet-architecture/eShopOnContainers/dev/src/Services/Ordering/Ordering.Domain/AggregatesModel/OrderAggregate/Order.cs))
there is the only way to manipulate units count in OrderItem - througth the agregate root, i.e. using  AddOrderItem in Order class.

And this is wery powerfull thing in tactical DDD, witch prevent a lot of error in my expirience.

But, AddUnits method in [OrderItem](https://raw.githubusercontent.com/dotnet-architecture/eShopOnContainers/dev/src/Services/Ordering/Ordering.Domain/AggregatesModel/OrderAggregate/OrderItem.cs) is public. and can be accidentally called directly and this may cause loss of aggregate consistency.

We can define AddUnits as internal, but then we need create assembly per aggregate and this approach not prevent all unwanted calls.

It would be great to have a friend access modifier. Its behavior should resemble the internal modifier (including combinations with protected), but not for the entire assembly, but only for the specified list of classes.

It might look something like this:

```
public class OrderItem: Entity
    friends: Order
{
    friend void AddUnits(int units)
    {
        if (units < 0)
        {
            throw new OrderingDomainException("Invalid units");
        }

        _units += units;
    }
}
```

Then we can use AddUnits only in Order class, what we wanted to achieve. 
A feature like this should not be difficult to implement, since it is similar to the internal modifier. And unlike similar functionality in C++, it does not violate encapsulation.

**Comment by HaloFour** [comment]
See: #2073

**Reply by K1vs** [reply]
I don’t want an implementation like in C++, which is criticized there. I suggest a more limited internal modifier variant.

**Reply by HaloFour** [reply]
The discussion goes into why the concept isn't really that desirable, regardless of the implementation.  You can achieve what you want with `internal` and an analyzer.

**Reply by K1vs** [reply]
I don’t see any real reason in that topic why this functionality might not be desirable for the language. In that topic there is well-deserved criticism and concerns about implementation in the C++ style, but for implementations similar to the one I proposed, there is no constructive criticism, but only advice on emulation. Yes, this functionality can be emulated using a code analyzer, but this statement is true for many features and syntactic sugar already present in the language. But nevertheless, it is much more convenient and reliable when this is a built-in language function that any developer is familiar with.

**Reply by HaloFour** [reply]
I think the discussion goes into detail as to why adding more accessibility levels like `friend` is undesirable.  Multiple C# team members have chimed in as to why they feel that it is not compelling.  I don't think slicing implementation details changes that equation any.

**Reply by CyrusNajmabadi** [reply]
My personal preference is htat instead of having the types declare friends, just allow types to be 'private' within a particular 'scope' of some sort.  Within that scope they can all play together, but not otherwise be exported outwards.

So similar to the 'file' concept for file-types, just not limited to a single file.

**Reply by K1vs** [reply]
CyrusNajmabadi, enumerating friendly types is one way to define this 'scope'. But nevertheless, a new access modifier is needed for limit access to  'only inside this scope'.

**Reply by DavidArno** [reply]
The way I'd like to see this type of situation resolved is by the `file` access modifier being supported for type members, not just for types. Both `Order` and `OrderItem` would be placed in the same file:
```cs
public class OrderItem: Entity
{
    file void AddUnits(int units) { ...
}

public class Order: Entity
{
    // AddUnits can be accessed from this class, but not by any other
}
```

**Reply by K1vs** [reply]
The access modifier 'file' is for code generation. And I would not really like to put all friendly classes in one file. This may not be convenient for long files. But even such a possibility would be better than what we have now.

**Reply by KennethHoff** [reply]
I think a namespace accessibility (as something between file and internal) would be nice.

```c#
namespace MyProject.Users;

public class User
{
    // .....
    namespace DoThing() { ...}
}

```

`DoThing()` would be accessible in `MyProject.Users` or lower (for example `MyProject.Users.Events`)

**Reply by theunrepentantgeek** [reply]
What do you do with namespaces that span assemblies? 

Namespace visibility isn't a strict sub-scope of internal.

**Reply by K1vs** [reply]
There are several problems with namespaces as an access scope in DotNet, unlike in Java, where packages are successfully used for this:

1) Hierarchy of namespaces
2) Cross assembly namespaces

If someone can come up with a good way to create logical access areas based on namespaces, I would be glad, but I don’t see it myself.

**Reply by KennethHoff** [reply]
You could have `internal namespace` and `public namespace` (presumably equivalent to simply `namespace`) which would work similarly to `private protected` and friends in that they'd work in conjunction with each other.

`internal namespace` would require same assembly and namespace. `public namespace` would simply require namespace

**Reply by HaloFour** [reply]
Java, until JDK9, also had no concept of modules.  The class hierarchy is just sprayed into a directory structure.  Java has no `internal` accessibility modifier as a result, although the equivalent in JMS would be `public` but not exported from the module.  This is why in JMS classes between two modules cannot exist in the same package.

Namespaces in C# are strictly that, part of the name of the type, and they only exist for organization and to help prevent naming collisions.  The runtime doesn't know or care anything about them.  All it knows is that `System.Collections.Generic.List` happens to have some dots in the name.

**Reply by DavidArno** [reply]
> Namespaces in C# are strictly that, part of the name of the type, and they only exist for organization and to help prevent naming collisions. The runtime doesn't know or care anything about them. All it knows is that System.Collections.Generic.List happens to have some dots in the name.

@HaloFour, this statement by you that eg `System.Console` is the type name at runtime with namespaces being an artifact of C# that doesn't exist at runtime sits at odds with [a statement made by @jnm2 almost two years ago](https://github.com/dotnet/csharplang/discussions/928#discussioncomment-1888899):
> Yes, `class A.B.C { }` would have to mean `class C { }` nested inside partial type `B` nested inside partial type `A`. I think @333fred or others were commenting in Discord a long while ago on how it would be unworkable for these to be namespaces.
>
> I do have a use case for `class Outer.Nested { }` related to what happens to source control line history and lack of rename history if you refactor a nested class to a top-level one or vice versa. I'm planning on drafting a proposal.

**Reply by HaloFour** [reply]
@DavidArno 

That's kind of a quirk of C#, which makes namespaces and nested classes appear to have the same name.  That's not the case in the runtime, where `/` is the separator between nested classes, e.g. `System.Console` vs. `Foo.Bar.Outer/Nested`.

**Reply by DavidArno** [reply]
@HaloFour ,

Does that `/` separator apply to all members of a type, or just subtypes? Trying to understand why `A.B.C` works in C# when `C` could be a type, a subtype of `B` or eg a property of `B`, but the runtime needs different notation.

**Reply by CyrusNajmabadi** [reply]
> Trying to understand why A.B.C works in C#

The language rules state what to do as each name is looked up.  For better or worse, our language uses the same member operator (`.`) for all these cases.

**Reply by HaloFour** [reply]
IL uses the `::` notation for members, so `[AssemblyName]Namespace.OuterClass/InnerClass::Method`.  I don't believe the runtime itself has any concept of qualifying members, though.  They only have a simple name and a declaring type.

It doesn't matter as far as C# is concerned, which has it's own name lookup rules.  However, it could bleed into any decisions about carving out new namespace-based accessibility modifiers, assuming that they're not just hacked in through a combination of `public`/`internal` accessibility and a `modreq`.

**Comment by Richiban** [comment]
I'm not commenting on your proposal, but on the example you gave.

I don't think the `AddUnits` method should exist at all (on OrderItem). I would expect to be able to simply add products repeatedly to the order, and only when the UI comes along does the list get grouped by product with a unit count.

**Reply by K1vs** [reply]
Specifically in this example, without delving into the business context, it is possible to model the orderItem as an immutable value object.  And I usually model the orderItem as a value object, if this corresponds to the understanding of its nature by business experts.  But in DDD there are quite frequent situations when an aggregate must have non-root entities.  By definition, these entities are mutable and have methods that affect their state.  And the language does not provide a way to allow these methods to be called exclusively from the aggregate class.

**Reply by HaloFour** [reply]
> And the language does not provide a way to allow these methods to be called exclusively from the aggregate class.

What languages do?  As far as I can tell, C++ is the only OOP language to provide anything like `friend` accessibility to create these hard dependencies between specific classes.  Other languages rely on modules and internal accessibility modifiers instead.  IMO, you shouldn't be creating such backdoor coupling between specific types that is so dangerous that the language should enforce that it can't even be called from within the same module, and I don't think that the language should encourage it either.  As the other discussion mentioned, there are several other ways to accomplish this goal that does not involve changing the language.  They may not be as ergonomic as a first-class language feature, but I don't think that adding new accessibility modifiers and scopes for every use case under the sun is a scalable or worthwhile endeavor.

**Reply by K1vs** [reply]
It is not a fundamental point to determine the scope of availability using the enumeration of classes. Even the namespace access modifier would be fine with me, or even the file access modifier mentioned above. But still, the friend would be the best option, as the most accurate way to define this access area. The problem with the internal modifier is that the assembly is not only a logical, but also a physical concept. And splitting the project into a huge number of assemblies for each aggregate will often be a bad decision. And although I don't see any problem in defining a list of friends of a class, an option with the possibility of assigning a class to a certain logical module and adding a 'module' acess modifier would also perfectly solve the problem.

```
module: Ordering;
public class OrderItem: Entity
{
    module void AddUnits(int units) { ...
}

module: Ordering;
public class Order: Entity
{
    // AddUnits can be accessed from this class, but not by any other
}
```

**Reply by K1vs** [reply]
> IMO, you shouldn't be creating such backdoor coupling between specific types that is so dangerous that the language should enforce that it can't even be called from within the same module, and I don't think that the language should encourage it either.

And although I do not agree with this statement, it is similarly true for internal modifier.

**Reply by HaloFour** [reply]
> And although I do not agree with this statement, it is similarly true for internal modifier.

I disagree, it is completely reasonable for an implementation to rely on components that are not a part of a public API.

**Reply by K1vs** [reply]
I understand what you mean. And here the main problem is the lack of a logical module smaller than the assembly and larger than the class.

**Reply by HaloFour** [reply]
At that point you're only trying to restrict access to the people working on that project, not to any consumers of an API or across module boundaries.  IMO that is a questionable need, one that can be satisfied through numerous other methods that don't involve changing the language and runtime to support a bunch more ways to slice and dice accessibility.

**Reply by K1vs** [reply]
Yes, this functionality is intended for the team that works with the project. But that doesn't make it unimportant. According to your logic, a private modifier is also not needed, because there is an internal, and a private one can be emulated by a code analyzer.

**Comment by colejohnson66** [comment]
I would assume the new “unsafe accessors” feature in .NET 8 would be Good Enough(tm) for such a situation. It’s not perfect, but it would work.

**Reply by K1vs** [reply]
This can be used as an ugly workaround, but not a solution to the problem.

**Reply by HaloFour** [reply]
Multiple solutions have been offered for this "problem".  Ultimately tho the problem lies in the design.  Trying to adhere to C++ design principles and porting directly to C# will result in friction, as would be the case in the other direction, or, frankly, between any two languages.  Stick to C# design principles and these kinds of problems largely go away.

IMO, this is very much a solution intended to be solved by analyzers.  It'd be trivial to use `internal` accessibility and define rules applied by attributes that specify allowlists between types, with any forms of policy you would want.  If there was enough demand for such a thing I would expect it to exist on NuGet already, although my limited attempt at a search has come up empty.  This would be the most expedient way to achieve your goals even if members of the language team expressed interest in it as a first-class feature.

**Reply by K1vs** [reply]
> Ultimately tho the problem lies in the design. Trying to adhere to C++ design principles and porting directly to C# will result in friction, as would be the case in the other direction, or, frankly, between any two languages. Stick to C# design principles and these kinds of problems largely go away.

I showed an example of a situation that occurs in a typical DDD case. And it is present in the example of building modern applications from Microsoft. This is not a consequence of C++ style design. In my example, this is dictated by the rules of tactical DDD, one of the most popular approaches to enterprise programming. This is a consequence of the lack of an intermediate size abstraction between the assembly and the class in DotNet. Yes, this problem is not terrible, and you can find many workarounds for it or simply ignore it. But I would like to have a solution to it, which is why I started the discussion.

**Reply by HaloFour** [reply]
> This is not a consequence of C++ style design.

Then why does it rely on a feature that is exclusive to C++?  No other language has this accessibility feature.  It's been deemed unnecessary in any language that has proper modularization.

One is expected to adapt the design patterns to the idioms of the target language.  Attempting to port one implementation directly from one language to another will only result in friction with the features, idioms and ecosystem of that language.

**Reply by K1vs** [reply]
> Then why does it rely on a feature that is exclusive to C++?

Forget about C++, I mentioned it only as an implementation example. The task of allowing access to a class member only for another class comes from tactical DDD. And if in C++ it can be solved very beautifully with friends, in Java it can be solved well via packages, but in C# it can only be solved clumsily with assemblies. Because it is not always rational to make a separate assembly for the unit, as the assembly is not only a logical boundary. Or a custom code analyzer, which is also not the best option. Or simply not solve this problem, relying on code reviews and the conscientiousness of the team.

**Reply by HaloFour** [reply]
> I mentioned it only as an implementation example.

I mentioned it because, so far, it's the only OOP language that I could find that has anything like `friend` accessibility.  If there were other languages that had other implementations that would make the conversation more interesting.

> Or a custom code analyzer, which is also not the best option.

Why not?  That is the entire point of analyzers.  It allows you to apply and enforce code policies for your projects.  It's an option that you could utilize immediately.

**Reply by K1vs** [reply]
> Why not? That is the entire point of analyzers. It allows you to apply and enforce code policies for your projects. It's an option that you could utilize immediately.

Because this is a custom solution, one team will have something of its own, another will have something else. And people coming to the project must learn this every time. However, I am currently working on such an analyzer, since access errors are periodically made by newcommers, and sometimes they even pass by the reviewer.

**Reply by HaloFour** [reply]
> Because this is a custom solution, one team will have something of its own, another will have something else. And people coming to the project must learn this every time. However, I am currently working on such an analyzer, since access errors are periodically made by newcommers, and sometimes they even pass by the reviewer.

One team has one set of policies, another team has nothing.  This is a virtue.  The language can't (and won't) scale to the infinite different ways that different groups of people want to slice the onion.

In a marketplace of ideas, an implementation that finds a traction across a wider audience could be deployed to NuGet and shared amongst the ecosystem.

**Comment by HaloFour** [comment]
Relevant:

https://github.com/dotnet/csharplang/blob/main/meetings/2023/LDM-2023-10-09.md#add-private-and-namespace-accessibility-modifiers-for-top-level-types

If `namespace` accessibility (or `private` for top-level types) are added that would effectively get you Java-like `package` accessibility.
